---
title: "fastq-to-counts"
author: "S2599932"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## This pipeline is to preprocess single cell RNA sequencing data generated using SMART-Seq 2 protocol

## Initial quality control using **FastQC**

```{bash}
for file in /home/s2599932/human-dataset/GSE36552_ENA_RAW/*.fastq
do
  fastqc -o /home/s2599932/human-dataset/GSE36552_ENA_RAW/fastqc-results "$file"
done
```

```{bash}
# Run multiqc to collate all the results
multiqc -o /home/s2599932/human-dataset/GSE36552_ENA_RAW/fastqc-results/multiqc-results /home/s2599932/human-dataset/GSE36552_ENA_RAW/fastqc-results/.
```

### Use **trimmomatic** to trim low quality reads

```{bash}
# Sample 1
java -jar ~/bin/Trimmomatic-0.39/trimmomatic-0.39.jar SE -phred33 /home/s2599932/human-dataset/GSE36552_ENA_RAW/SRR491070.fastq /home/s2599932/human-dataset/GSE36552_ENA_RAW/trimmed-fastq/SRR491070_trimmed.fastq LEADING:20 TRAILING:20 SLIDINGWINDOW:4:20 MINLEN:36
```

```{bash}
# Sample 2
java -jar ~/bin/Trimmomatic-0.39/trimmomatic-0.39.jar SE -phred33 /home/s2599932/human-dataset/GSE36552_ENA_RAW/SRR490970.fastq /home/s2599932/human-dataset/GSE36552_ENA_RAW/trimmed-fastq/SRR490970_trimmed.fastq LEADING:20 TRAILING:20 SLIDINGWINDOW:4:20 MINLEN:36
```

### Run **FastQC** on the trimmed fastq files

```{bash}
for file in /home/s2599932/human-dataset/GSE36552_ENA_RAW/trimmed-fastq/*.fastq
do
  fastqc -o /home/s2599932/human-dataset/GSE36552_ENA_RAW/trimmed-fastq/trimmed-fastqc-results/ "$file"
done
```

### Run **MultiQC** to collate all the results

```{bash}
multiqc -o /home/s2599932/human-dataset/GSE36552_ENA_RAW/trimmed-fastq/trimmed-fastqc-results/trimmed-multiqc /home/s2599932/human-dataset/GSE36552_ENA_RAW/trimmed-fastq/trimmed-fastqc-results.
```

### Use **STAR** to align the reads to reference genome

```{bash}
# Create the reference genome after downloading the human genome fasta file and annotation .gtf file
STAR --runThreadN 10 \
     --runMode genomeGenerate \
     --genomeDir /home/s2599932/pre-processing/alignment/ref \
     --genomeFastaFiles /home/s2599932/pre-processing/alignment/fasta/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa \
     --sjdbGTFfile /home/s2599932/pre-processing/alignment/fasta/Homo_sapiens.GRCh38.112.gtf \
     --sjdbOverhang 99
```

```{bash}
# Define input and output directories
input_dir="/home/s2599932/human-dataset/GSE36552_ENA_RAW/trimmed-fastq"
output_dir="/home/s2599932/pre-processing/alignment/mapped"
genome_dir="/home/s2599932/pre-processing/alignment/ref"

# Loop over each FASTQ file in the input directory
for fastq_file in ${input_dir}/*.fastq
do
  # Extract the base name of the file (without path and extension)
  base_name=$(basename ${fastq_file} _trimmed.fastq)
  
  # Define the output prefix for this file
  output_prefix="${output_dir}/${base_name}_"

  # Run STAR for the current file
  STAR --runThreadN 12 \
       --genomeDir ${genome_dir} \
       --readFilesIn ${fastq_file} \
       --outFileNamePrefix ${output_prefix} \
       --outSAMtype BAM SortedByCoordinate
done
```

### Check the quality of alignment using **samtools**

```{bash}
# Define the output directory where BAM files are located
output_dir="/home/s2599932/pre-processing/alignment/mapped"

# Loop over each BAM file in the output directory
for bam_file in ${output_dir}/*_Aligned.sortedByCoord.out.bam
do
  # Extract the base name of the file (without path and extension)
  base_name=$(basename ${bam_file} _Aligned.sortedByCoord.out.bam)

  # Define the output file for flagstat results
  flagstat_output="${output_dir}/${base_name}_flagstat.txt"

  # Run samtools flagstat for the current BAM file
  samtools flagstat ${bam_file} > ${flagstat_output}
done
```

### Use **featureCounts** to quantify the aligned reads

> Will generate individual counts file for each bam file

```{bash}
# Define directories and annotation file
bam_dir="/home/s2599932/pre-processing/alignment/mapped"
gtf_file="/home/s2599932/pre-processing/alignment/fasta/Homo_sapiens.GRCh38.112.gtf"
output_dir="/home/s2599932/pre-processing/alignment/counts"

# Create output directory if it doesn't exist
mkdir -p ${output_dir}

# Loop over each BAM file in the bam directory
for bam_file in ${bam_dir}/*_Aligned.sortedByCoord.out.bam
do
  # Extract the base name of the file (without path and extension)
  base_name=$(basename ${bam_file} _Aligned.sortedByCoord.out.bam)

  # Define the output file for featureCounts results
  counts_output="${output_dir}/${base_name}_featureCounts.txt"

  # Run featureCounts for the current BAM file
  featureCounts -T 12 \
                -a ${gtf_file} \
                -o ${counts_output} \
                ${bam_file}
done
```

> Will generate a single compiled counts file, thats what we need!!

```{bash}
# Define directories and annotation file
bam_dir="/home/s2599932/pre-processing/alignment/mapped"
gtf_file="/home/s2599932/pre-processing/alignment/fasta/Homo_sapiens.GRCh38.112.gtf"
output_dir="/home/s2599932/pre-processing/alignment/counts"

# Create output directory if it doesn't exist
mkdir -p ${output_dir}

# Collect all BAM files into a single variable
bam_files=$(ls ${bam_dir}/*_Aligned.sortedByCoord.out.bam)

# Define the output file for featureCounts results
counts_output="${output_dir}/all_samples_featureCounts.txt"

# Run featureCounts for all BAM files
featureCounts -T 12 \
              -a ${gtf_file} \
              -o ${counts_output} \
              ${bam_files}
```


### The all_sample_featureCounts.txt must have the count data for every sample along with some additional columns that has to dropped in the downstream analysis.
















